cloudify/amqp_client.py

```diff
@@ -263,7 +276,46 @@ def close(self, wait=True):
     def add_handler(self, handler):
         self._handlers.append(handler)
         if self._pika_connection:
-            handler.register(self._pika_connection, self._publish_queue)
+            handler.register(self)
+
+    def channel(self):
+        if self._closed or not self._pika_connection:
+            raise RuntimeError(
+                'Attempted to open a channel on a closed connection')
+        return self._pika_connection.channel()
+
+    def publish(self, message, wait=True):
+        """Schedule a message to be sent.
+
+        :param message: Kwargs for the pika basic_publish call. Should at
+                        least contain the "body" and "exchange" keys, and
+                        it might contain other keys such as "routing_key"
+                        or "properties"
+        :param wait: Whether to wait for the message to actually be sent.
+                     If true, an exception will be raised if the message
+                     cannot be sent.
+        """
+        if wait and self._consumer_thread \
+                and self._consumer_thread is threading.current_thread():
+            # when sending from the connection thread, we can't wait because
+            # then we wouldn't allow the actual send loop (._process_publish)
+            # to run, because we'd block on the err_queue here
+            raise RuntimeError(
+                'Cannot wait when sending from the connection thread')
+
+        # the message will likely be sent from another thread (the .consume
+        # thread). If an error happens there, we must have a way to get it
+        # back out, so we pass a Queue together with the message, that will
+        # contain either an exception instance, or None
+        err_queue = Queue.Queue() if wait else None
+        envelope = {
+            'message': message,
+            'err_queue': err_queue
+        }
+        self._publish_queue.put(envelope)
+        err = err_queue.get()
```

Comment:
Also, do we want it to block forever, or do we want a timeout here? Can there be an unhandled scenario on the other side, that will cause this thread to block forever?